tools/compare_pages.sh

```diff
@@ -1,29 +1,135 @@
 #!/usr/bin/env bash
 
-FILE_TO_COMPARE=$1
-LANGUAGE=$2
-FIRST_PAGE=$3
-
-if [[ -z $4 ]]; then
-  LAST_PAGE=$3
-else
-  LAST_PAGE=$4
+cache_dir="$(pwd)/cache"
+
+#
+# HELPER FUNCTIONS
+#
+
+help() {
+  echo "Usage: $0 -l <language> [--language=<language>] [-p] [--printable] -r <range> [--range=<range>]"
+  echo "-r, --range: use single page number or dash-separated range of pages"
+  exit 2
+}
+
+url() {
+  type=$([[ "$2" -eq 1 ]] && echo "printable" || echo "main")
+  echo "https://raw.githubusercontent.com/qwrtln/Homm3BG-build-artifacts/${1}/${type}_${1}.pdf"
+}
+
+file_type() {
+  [[ "$1" -eq 1 ]] && echo "printable" || echo "main"
+}
+
+base_file_path() {
+  local language="$1"
+  local printable="$2"
+  type=$(file_type $printable)
+
+  echo "${cache_dir}/${type}_${language}.pdf"
+}
+
+download_base_file() {
+  echo "Downloading base file..."
+
+  local language="$1"
+  local printable="$2"
+  type=$(file_type "$printable")
+
+  url="https://raw.githubusercontent.com/qwrtln/Homm3BG-build-artifacts/${language}/${type}_${language}.pdf"
+  curl -o "${cache_dir}/${type}_${language}.pdf" "$url"
+}
+
+file_mod_time() {
+  local file=$1
+  if [[ "$(uname)" == "Darwin" ]]; then
+    stat -f %m "$file"
+  else 
+    stat -c %Y "$file"
+  fi
+}
+
+ensure_base_file() {
+  echo "Checking if there is the base file for comparison..."
+
+  local language="$1"
+  local printable="$2"
+  local base_file=$(base_file_path "$language" "$printable")
+
+  if [[ -f $base_file ]]; then
+    mod_time=$(file_mod_time "$base_file")
+    now=$(date +%s)
+    age=$((now - mod_time))  # seconds
+
+    if [[ $age -gt 3600 ]]; then
+      download_base_file "$language" "$printable"
+    fi
+  else
+    download_base_file "$language" "$printable"
+  fi
+}
+
+#
+# MAIN FLOW
+#
+
+language=""
+printable=0
+range=""
+
+while [[ "$1" != "" ]]; do
+  case $1 in
+    -l | --language )
+      shift
+      language=$1
+      ;;
+    -p | --printable )
+      printable=1
+      ;;
+    -r | --range )
+      shift
+      range=$1
+      ;;
+    * )
+      help
+      ;;
+  esac
+  shift
+done
+
+if [[ -z "$language" || -z "$range" ]]; then
+  help
+fi
+
+if [[ ! -d "$cache_dir" ]]; then
+  mkdir $cache_dir
+fi
+
```

Comment:
```suggestion
mkdir -p $cache_dir
```

Reply:
It only creates 1 level, `-p` would have no effect here

Reply:
smart